util/grub-set-bootflag.1
.TH GRUB-SET-BOOTFLAG 1 "Tue Jun 12 2018"
.SH NAME
\fBgrub-set-bootflag\fR \(em Set a bootflag in the GRUB environment block.

.SH SYNOPSIS
\fBgrub-set-bootflag\fR <\fIBOOTFLAG\fR>

.SH DESCRIPTION
\fBgrub-set-bootflag\fR is a command line to set bootflags in GRUB's
stored environment.

.SH COMMANDS
.TP
\fBBOOTFLAG\fR
.RS 7
Bootflag to set, one of \fIboot_success\fR or \fImenu_show_once\fR.
.RE

.SH SEE ALSO
.BR "info grub"


NEWS
New in 2.02:

* New/improved filesystem and disk support:
  * Big-endian UFS1.
  * Experimental 64-bit ext2 support.
  * Various fixes for non-512-byte sector devices.
  * New `proc' filesystem framework, used by LUKS disks.
  * Fix DM-RAID partition handling.
  * New `nativedisk' command to switch from firmware to native disk drivers.
  * Compressed HFS+.
  * DragonFly BSD labels.
  * CBFS (coreboot).
  * Handle partitioned LVM properly.
  * Use LVM UUIDs whenever possible.
  * GPT PReP.
  * New `progress' module that shows progress information while reading
    files.
  * ZFS features support.
  * ZFS LZ4 support.

* New/improved terminal and video support:
  * Monochrome text (matching `hercules' in GRUB Legacy).
  * Morse code output using system speaker.
  * `spkmodem' output (simple data protocol using system speaker).
  * Handle Japanese special keys.
  * coreboot framebuffer.
  * Serial on ARC.
  * Native vt100 handling for grub-emu, replacing the use of the curses
    library.
  * New gfxmenu options for terminal window positioning, theme background
    image handling, and scrollbar padding, plus `item_pixmap_style' and
    `highlight_overlay'.
  * Support several more image types (paletted and greyscale).

* Boot protocol improvements:
  * Support Apple FAT binaries on non-Apple platforms.
  * Improve FreeDOS direct loading support compatibility.
  * Enable `linux16' on all x86 platforms, not just BIOS.
  * New TrueCrypt ISO loader.
  * multiboot2 boot-services EFI specification.
  * multiboot2 EFI memory map specification.
  * multiboot2 full-file specfication.

* New/improved network support:
  * New variables `net_default_*' containing properties of the default
    interface.
  * Autoload `http' and `tftp' modules if necessary.
  * Improve TFTP robustness.
  * Parse `nd' disk names in GRUB Legacy configuration files.
  * Issue separate DNS queries for IPv4 and IPv6.

* Coreboot improvements:
  * CBFS support both in on-disk images (loopback) and flash.
  * Ability to launch another payload from flash or disk
  * Coreboot framebuffer
  * CBMEMC support (both logging and inspecting logs)
  * Command for inspecting coreboot timestamps (`coreboot_boottime').
  * Command for inspecting coreboot tables (`lscoreboot').
  * New target default_payload.elf.
  * Increased maximal core size.

* New/improved platform support:
  * New `efifwsetup' and `lsefi' commands on EFI platforms.
  * New `cmosdump' and `cmosset' commands on platforms with CMOS support.
  * New command `pcidump' for PCI platforms.
  * New `cbmemc', `coreboot_boottime', and `lscoreboot' commands on
    coreboot.
  * Improve opcode parsing in ACPI halt implementation.
  * Use the TSC as a possible time source on i386-ieee1275.
  * Merge PowerPC grub-mkrescue implementation with the common one.
  * Support grub-mkrescue on i386-ieee1275, sparc64, bootinfo machines such
    as pSeries, and mips-arc.
  * Make grub-mkrescue better support Apple Intel Macs on CD.
  * Enable GRUB Legacy configuration file parsing on EFI.
  * Support halt for Loongson 2E.
  * ARM U-Boot and EFI ports.
  * Reorganise platform-dependent code in utilities to avoid #ifdef mess.
  * AROS and Haiku support for userspace utilities.
  * Xen PV port.
  * Fix EFI stack alignment.
  * ARM64 EFI port.
  * On Linux, read partition start offsets from sysfs if possible.
  * New grub-macbless utility, and better integration with Mac firmware in
    grub-install.
  * Support Yeeloong 3A.
  * Add `cpuid --pae' option to detect Physical Address Extension on x86.
  * Support for USB debug dongles.
  * Support for *-emu on all platforms (previously only i386/x86_64 worked).
  * Support *-emu on Windows.

* Security:
  * Add optional facility to enforce that all files read by the core image
    from disk have a valid detached digital signature.

* Performance:
  * Avoid costly division operations in many places.
  * New boot time analysis framework (`./configure --enable-boot-time').
  * Initialise USB ports in parallel.
  * New `testspeed' command to test file read speed.
  * Speed-up gfxterm by storing intermediate results in more compact format.
  * Lazy LVM/mdraid scan.
  * Disk hints.

* Scripting:
  * New `eval' and `tr' commands.
  * grub-script-check fails on scripts containing no commands.

* Installation and other utility improvements:
  * Add option to compress files on installation or image creation.
  * Using grub-reboot no longer requires setting `GRUB_DEFAULT=saved'.
  * Support probing EFI System Partition (requires os-prober >= 1.58).
  * Fix inconsistent use of `GRUB_CRYPTODISK_ENABLE' and
    `GRUB_ENABLE_CRYPTODISK'; the latter is now used consistently.
  * grub-mount handles symbolic links to directories.
  * Support disabling submenus with `GRUB_DISABLE_SUBMENU' configuration key
    for grub-mkconfig.
  * grub-install, grub-mknetdir, grub-mkrescue, and grub-mkstandalone
    rewritten in C.  They should now work in supported non-Unix-like
    environments.
  * Native mingw support.
  * Ability to install on EFI under windows.
  * Reorganise timeout handling using new `timeout_style' environment
    variable and `GRUB_TIMEOUT_STYLE' configuration key for grub-mkconfig.
    Menu hotkeys pressed during a hidden timeout now boot the corresponding
    menu entry immediately.
  * New `file' command and grub-file utility to check file types.
  * New syslinux configuration file parser.

* Build system:
  * Remove all uses of nested functions; GRUB no longer requires an
    executable stack.
  * Fix documentation build with Texinfo >= 5.1.
  * More robust and documented cross-compiling support.
  * Partial clang support for some platforms (experimental).
  * Partial mingw64 x86_64-efi compile support (highly experimental).
  * Partial mingw32 i386-* (other than already present i386-pc)
    compile support (highly experimental).
  * Support for grub-mkpasswd on Windows.
  * Eliminate the use of AutoGen.  This allowed some performance

util/grub-set-bootflag.c
/* grub-set-bootflag.c - tool to set boot-flags in the grubenv. */
/*
 *  GRUB  --  GRand Unified Bootloader
 *  Copyright (C) 2018 Free Software Foundation, Inc.
 *
 *  GRUB is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  GRUB is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * NOTE this gets run by users as root (through pkexec), so this does not
 * use any grub library / util functions to allow for easy auditing.
 * The grub headers are only included to get certain defines.
 */

#include <config-util.h>     /* For *_DIR_NAME defines */
#include <grub/types.h>
#include <grub/lib/envblk.h> /* For GRUB_ENVBLK_DEFCFG define */
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "progname.h"

#define GRUBENV "/" GRUB_BOOT_DIR_NAME "/" GRUB_DIR_NAME "/" GRUB_ENVBLK_DEFCFG
#define GRUBENV_SIZE 1024

const char *bootflags[] = {
  "boot_success",
  "menu_show_once",
  NULL
};

static void usage(FILE *out)
{
  int i;

  fprintf (out, "Usage: 'grub-set-bootflag <bootflag>', where <bootflag> is one of:\n");
  for (i = 0; bootflags[i]; i++)
    fprintf (out, "  %s\n", bootflags[i]);
}

int main(int argc, char *argv[])
{
  /* NOTE buf must be at least the longest bootflag length + 4 bytes */
  char env[GRUBENV_SIZE + 1], buf[64], *s;
  const char *bootflag;
  int i, len, ret;
  FILE *f;

  if (argc != 2)
    {
      usage (stderr);
      return 1;
    }
  else if (!strcmp (argv[1], "--help"))
    {
      usage (stdout);
      return 0;
    }
  else if (!strcmp (argv[1], "--version"))
    {
      printf ("grub-set-bootflag (%s) %s\n", PACKAGE_NAME, PACKAGE_VERSION);
      return 0;
    }

  for (i = 0; bootflags[i]; i++)
    if (!strcmp (argv[1], bootflags[i]))
      break;
  if (!bootflags[i])
    {
      fprintf (stderr, "Invalid bootflag: '%s'\n", argv[1]);
      usage (stderr);
      return 1;
    }

  bootflag = bootflags[i];
  len = strlen (bootflag);

  f = fopen (GRUBENV, "r");
  if (!f)
    {
      perror ("Error opening " GRUBENV " for reading");
      return 1;     
    }

  ret = fread (env, 1, GRUBENV_SIZE, f);
  fclose (f);
  if (ret != GRUBENV_SIZE)
    {
      errno = EINVAL;
      perror ("Error reading from " GRUBENV);
      return 1;     
    }

  /* 0 terminate env */
  env[GRUBENV_SIZE] = 0;

  if (strncmp (env, GRUB_ENVBLK_SIGNATURE, strlen (GRUB_ENVBLK_SIGNATURE)))
    {
      fprintf (stderr, "Error invalid environment block\n");
      return 1;
    }

  /* Find a pre-existing definition of the bootflag */
  s = strstr (env, bootflag);
  while (s && s[len] != '=')
    s = strstr (s + len, bootflag);

  if (s && ((s[len + 1] != '0' && s[len + 1] != '1') || s[len + 2] != '\n'))
    {
      fprintf (stderr, "Pre-existing bootflag '%s' has unexpected value\n", bootflag);
      return 1;     
    }

  /* No pre-existing bootflag? -> find free space */
  if (!s)
    {
      for (i = 0; i < (len + 3); i++)
        buf[i] = '#';
      buf[i] = 0;
      s = strstr (env, buf);
    }

  if (!s)
    {
      fprintf (stderr, "No space in grubenv to store bootflag '%s'\n", bootflag);
      return 1;     
    }

  /* The grubenv is not 0 terminated, so memcpy the name + '=' , '1', '\n' */
  snprintf(buf, sizeof(buf), "%s=1\n", bootflag);
  memcpy(s, buf, len + 3);

  /* "r+", don't truncate so that the diskspace stays reserved */
  f = fopen (GRUBENV, "r+");
  if (!f)
    {
      perror ("Error opening " GRUBENV " for writing");
      return 1;     
    }

  ret = fwrite (env, 1, GRUBENV_SIZE, f);
  if (ret != GRUBENV_SIZE)
    {
      perror ("Error writing to " GRUBENV);
      return 1;     
    }

  ret = fflush (f);
  if (ret)
    {
      perror ("Error flushing " GRUBENV);
      return 1;     
    }

  fsync (fileno (f));
  fclose (f);

  return 0;
}
